### PR TITLE
heartbeat and stat sync fixes

### PR DESCRIPTION
**Pull request by kinarashah**

### CODE REVIEW

File: controller/clusterheartbeat/heartbeatsyncer.go [hunk 1]
```diff
@@ -15,53 +18,92 @@ const (
 	syncInterval = 20 * time.Second
 )
 
-var clusterToLastUpdated map[string]time.Time
+type updateTimes struct {
+	old time.Time
+	new time.Time
```
Comment (alena1108) [inline controller/clusterheartbeat/heartbeatsyncer.go]: probably should not use the `new` keyword as it is reserved

File: controller/clusterheartbeat/heartbeatsyncer.go [hunk 2]
```diff
@@ -15,53 +18,92 @@ const (
...
+	mapLock              = sync.Mutex{}
+)
 
 type HeartBeatSyncer struct {
 	ClusterClient v3.ClusterInterface
 }
 
-func Register(management *config.ManagementContext) {
+type MachineSyncer struct {
+	MachineClient v3.MachineInterface
```
Comment (alena1108) [inline controller/clusterheartbeat/heartbeatsyncer.go]: no need to have machineClient as a field on the MachineSyncer. You should just add lifecycle as this directly in Register:
`management.Management.Machines("").AddLifecycle`

File: controller/clusterheartbeat/heartbeatsyncer.go [hunk 3]
```diff
@@ -15,53 +18,92 @@ const (
...
-		}
-	} else {
-		condition := getConditionIfReady(cluster)
-		if condition != nil {
-			lastUpdateTime, _ := time.Parse(time.RFC3339, condition.LastUpdateTime)
-			clusterToLastUpdated[key] = lastUpdateTime
-			logrus.Debugf("Synced cluster [%s] successfully", key)
+func (h *HeartBeatSyncer) Create(cluster *v3.Cluster) (*v3.Cluster, error) {
+	h.updateLastUpdateTime(cluster.Name, cluster)
+	return cluster, nil
```
Comment (alena1108) [inline controller/clusterheartbeat/heartbeatsyncer.go]: We should only return an object if something got modified, and that object should be the one with the updates, not the original one that was passed in. So it is either nil, nil when nothing changed, or updated, nil when something changed. Same for updated/removed

File: controller/clusterheartbeat/heartbeatsyncer.go [hunk 4]
```diff
@@ -15,53 +18,92 @@ const (
...
+}
+
+func (h *HeartBeatSyncer) Updated(cluster *v3.Cluster) (*v3.Cluster, error) {
+	h.updateLastUpdateTime(cluster.Name, cluster)
```
Comment (alena1108) [inline controller/clusterheartbeat/heartbeatsyncer.go]: nit: there is no need for cluster.Name in the method signature. Just derive it from cluster in the underlying call

File: controller/clusterheartbeat/heartbeatsyncer.go [hunk 5]
```diff
@@ -102,3 +134,28 @@ func getConditionIfReady(cluster *v3.Cluster) *v3.ClusterCondition {
...
+	return m.sync(machine)
+}
+
+func (m *MachineSyncer) Remove(machine *v3.Machine) (*v3.Machine, error) {
+	return m.sync(machine)
+}
+
+func (m *MachineSyncer) sync(machine *v3.Machine) (*v3.Machine, error) {
+	m.Clusters.Controller().Enqueue("", machine.ClusterName)
+	return machine, nil
```
Comment (alena1108) [inline controller/clusterheartbeat/heartbeatsyncer.go]: nothing changed on the machine, so return nil, nil

File: controller/clusterstats/statsaggregator.go [hunk 6]
```diff
@@ -3,14 +3,14 @@ package clusterstats
...
-	"github.com/sirupsen/logrus"
 	"k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/resource"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
 )
 
 type StatsAggregator struct {
 	Clusters v3.ClusterInterface
+	Machines v3.MachineLister
```
Comment (alena1108) [inline controller/clusterstats/statsaggregator.go]: you should use the cache here: v3.MachineLister, so the machine objects are fetched from cache, not from the api directly

File: controller/clusterstats/statsaggregator.go [hunk 7]
```diff
@@ -3,14 +3,14 @@ package clusterstats
...
 	"github.com/rancher/types/config"
-	"github.com/sirupsen/logrus"
 	"k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/resource"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
 )
 
 type StatsAggregator struct {
 	Clusters v3.ClusterInterface
```
Comment (alena1108) [inline controller/clusterstats/statsaggregator.go]: there is no need for clusters to be defined as a field on StatsAggregator. Just add a lifecycle directly to it in Register.

File: controller/clusterstats/statsaggregator.go [hunk 8]
```diff
@@ -112,40 +69,40 @@ func (s *StatsAggregator) aggregate(cluster *v3.Cluster, clusterName string) {
 	condDisk := v1.ConditionTrue
 	condMem := v1.ConditionTrue
 
-	for _, v := range stats[clusterName] {
-		if v == nil {
+	for _, machine := range machines {
+		if machine == nil || clusterName != machine.Status.ClusterName {
```
Comment (alena1108) [inline controller/clusterstats/statsaggregator.go]: machine can't be nil here

File: controller/clusterstats/statsaggregator.go (hunk 6 above)
Comment (kinarashah) [reply controller/clusterstats/statsaggregator.go]: I think I am using the cache for Machines as
```
Machines: management.Management.Machines("").Controller().Lister(),
```

File: controller/clusterheartbeat/heartbeatsyncer.go [hunk 9]
```diff
@@ -1,89 +1,137 @@
...
+	mapLock              = sync.Mutex{}
+)
 
 type HeartBeatSyncer struct {
-	ClusterClient v3.ClusterInterface
+	ClusterLister v3.ClusterLister
 }
 
-func Register(management *config.ManagementContext) {
+type MachineSyncer struct {
```
Comment (alena1108) [inline controller/clusterheartbeat/heartbeatsyncer.go]: @kinarashah I'd rather make this syncer a part of statsaggregator.go, as functionality wise it belongs there, and has really nothing to do with the heartbeat

**Pull request merged**